README.md
# Assume IAM Roles using the Hub-Spoke model

When you have _several_ AWS accounts to manage, you can keep things secure and locked-down by adopting the hub-spoke model of assuming IAM roles across accounts.

1. You have a user (or a bot if you're automating) with permission to assume the "Hub" role in an account (doesn't need to be the same as the user).

1. From the "Hub", the user can then traverse to a "Spoke" account to perform the actions that are granted to an assumer of that "Hub" role.

This model is recommended by AWS (read below). You will need to provision the roles (via [Service Control Policies](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_scps.html) or perhaps [Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role)), but there's only one user (per workload) that you need to grant permissions to.

## What is the Hub-Spoke model?

In a multi-account setup, optionally managed with AWS Organizations and AWS SSO, think of it like a bicycle wheel: One hub, many spokes.

For example, if an automated process (e.g., a robot) needs to perform the same kinds of actions in every account you own (e.g., security analysis, reporting, account inventory), you would set up:

1. A "service account" for the robot/job/process/whatever.
1. Designate one account as the "hub". You need to connect here before connecting to anything else. It is conceptually similar to a "jump box".
1. All accounts where actions need to be performed are the "spokes". They would all have the same policy that can be assumed by the user connecting through the hub account and out to spoke accounts.

In larger setups that use AWS Control Tower, these policies can be provisioned with _Service Control Policies_ (SCPs). In smaller setups, you can use tools like Terraform or the AWS CLI for automation.

## Install as a CLI tool

1. You must have the Golang toolchain installed first.

    ```bash
    brew install go
    ```

1. Add `$GOPATH/bin` to your `$PATH` environment variable. By default (i.e., without configuration), `$GOPATH` is defined as `$HOME/go`.

    ```bash
    export PATH="$PATH:$GOPATH/bin"
    ```

1. Once you've done everything above, you can use `go get`.

    ```bash
    go get github.com/northwood-labs/assume-spoke-role
    ```

## Usage as CLI Tool

```bash
# Learn how it works.
assume-spoke-role --help
```

Run a command in another account (assuming you have permissions to assume a role). The ` -- ` marker signifies the end of passing options, and to begin treating subsequent text as the command to run with those credentials.

Assuming you're using [AWS Vault](https://github.com/99designs/aws-vault) to manage your credentials, and want to manage common configurations via environment variables:

```bash
export ASSUME_ROLE_EXTERNAL_ID=this-is-my-robot
export ASSUME_ROLE_HUB_ACCOUNT=999999999999
export ASSUME_ROLE_HUB_ROLE=robot-hub-role
export ASSUME_ROLE_SPOKE_ROLE=robot-spoke-role

# Using your local credentials (sys-robot), assume a role in the "HUB" account, before
# pivoting to a "SPOKE" account, then executing a command with those "SPOKE"
# credentials.
aws-vault exec sys-robot --no-session -- \
    assume-spoke-role --spoke-account 888888888888 -- \
        aws sts get-caller-identity
```

Or, if you want to more explicitly rely on CLI parameters rather than environment variables:

```bash
aws-vault exec sys-robot --no-session -- \
    assume-spoke-role \
        --hub-account 999999999999 \
        --spoke-account 888888888888 \
        --hub-role robot-hub-role \
        --spoke-role robot-spoke-role \
        --external-id this-is-my-robot \
        -- \
            aws sts get-caller-identity
```

## Usage as Library

This can also be used as a library in your own applications for generating a set of STS credentials.

```go
import (
    "github.com/northwood-labs/assume-spoke-role/hubspoke"
	"github.com/northwood-labs/awsutils"
)

func main() {
    // Get AWS credentials from environment.
    config, err := awsutils.GetAWSConfig(ctx, "", "", 3, false)
    if err != nil {
        log.Fatal(fmt.Sprintf("could not generate a valid AWS configuration object: %w", err))
    }

    // Assume appropriate roles and return session credentials for the "Spoke" account.
    roleCredentials, _, err := hubspoke.GetSpokeCredentials(&hubspoke.SpokeCredentialsInput{
        Context:        ctx,
        Config:         &config,
        HubAccountID:   "888888888888",
        SpokeAccountID: "999999999999",
        HubRoleName:    "hub-role",
        SpokeRoleName:  "spoke-role",
        ExternalID:     "abc123",
        SessionString:  "me@email.com",
    })
    if err != nil {
        log.Fatal(fmt.Sprintf("could not generate valid AWS credentials for the 'spoke' account: %w", err))
    }

    fmt.Printf("AWS_ACCESS_KEY_ID=%s\n", *roleCredentials.AccessKeyId),
    fmt.Printf("AWS_SECRET_ACCESS_KEY=%s\n", *roleCredentials.SecretAccessKey),
    fmt.Printf("AWS_SECURITY_TOKEN=%s\n", *roleCredentials.SessionToken),
    fmt.Printf("AWS_SESSION_TOKEN=%s\n", *roleCredentials.SessionToken),
    fmt.Printf("AWS_SESSION_EXPIRATION=%s\n", roleCredentials.Expiration.String()),
}
```

See `cmd_run.go`, which implements this library to produce this very same CLI tool.

## Setting up the Hub/Spoke configuration

Following the [Principle of Least Privilege](https://www.cisecurity.org/spotlight/ei-isac-cybersecurity-spotlight-principle-of-least-privilege/), we're going to scope-down the permissions to as few as necessary.

### The User

In one of your AWS accounts, create an IAM user/instance-profile/whatever dedicated to this task. Since this user represents a _process_ and not a _person_, I recommend prefixing the user name with `sys-`. If we wanted to do things on behalf of the "robot" process, then perhaps we'd call this user `sys-robot`.

Just like a [Meeseeks](https://rickandmorty.fandom.com/wiki/Mr._Meeseeks), this user is intended for only a single task. It's better to have more users (with corresponding spoke roles) with fewer permissions, than it is to have fewer users (with corresponding spoke roles) with more permissions. Please don't re-use the same user for many tasks, as you increase your cybersecurity "blast radius" that way.

This user — as itself — can only do one thing: assume an IAM role in the "hub" account. (Replace `{hub-account-id}` with the AWS Account ID where your "hub" role is located.)

**Policy:**

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "sts:AssumeRole",
      "Resource": [
        "arn:aws:iam::{hub-account-id}:role/robot-hub-role"
      ]
    }
  ]
}
```

#### AWS Vault

As an enlightened user of AWS, you use [AWS Vault](https://github.com/99designs/aws-vault) (or maybe [AWS Okta](https://github.com/fiveai/aws-okta)) to manage your credentials. This stores them in the system keychain instead of as plain text on-disk, it automatically generates STS session credentials on your behalf, and it's easy to pass the credentials to things that are built with the AWS SDKs _besides_ the AWS CLI. Oh — and it also supports AWS SSO out-of-the-box.

### The Hub

Using the "robot" process example, let's follow-through with creating an IAM role to assume, and call it `robot-hub-role`.

This is an IAM role which will grant access to your user for assuming the "spoke" role in every account which has that identically-named role.

(If you're not using AWS Organizations, you can remove the entire `Condition` block. You should also specify the AWS Account IDs of the accounts you want to access.)

**Policy:**

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "sts:AssumeRole"
      ],
      "Resource": "arn:aws:iam::*:role/robot-spoke-role",
      "Condition": {
        "StringEquals": {
          "aws:PrincipalOrgID": "o-ZZZZZZZZZZ"
        }
      }
    }
  ]
}

```

You'll also need to configure the _trust relationship_ for the "hub" role so that only our user can assume it.

(If you're not using AWS Organizations, you can remove the entire `Condition` block.)

**Trust Relationship:**

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::{user-creation-account-id}:user/sys-robot"
      },
      "Action": "sts:AssumeRole",
      "Condition": {
        "StringEquals": {
          "aws:PrincipalOrgID": "o-ZZZZZZZZZZ"
        }
      }
    }
  ]
}
```

This creates a bi-directional symbiosis where the user can only assume the hub role, and the hub role can only be assumed by the user.

### The Spoke

Using the "robot" process example, let's follow-through with creating an IAM role to assume, and call it `robot-spoke-role`.

This is an IAM role which will grant access to your user (via the hub role) and grants the permissions for what can be done in this account. In our case, we want to grant `ReadOnlyAccess` (the built-in, AWS managed policy). Your needs may be different, so adapt accordingly.

You will need to configure the _trust relationship_ for the "spoke" role so that only our "hub" role can access it.

For an extra bit of entropy in our security, we can require an _External ID_ which is known only to the IAM role and the user accessing it.

**Policy:**

This should be a policy which lists the things that the assuming user is permitted to do in the spoke account.

**Trust Relationship:**

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::{hub-account-id}:role/robot-hub-role"
      },
      "Action": "sts:AssumeRole",
      "Condition": {
        "StringEquals": {
          "sts:ExternalId": "{your-external-id}"
        }
      }
    }
  ]
}
```

This creates a bi-directional symbiosis where the hub role can only assume the spoke role, and the spoke role can only be assumed by the hub role.

## Development

### Requirements

TBD

### Testing

TBD


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 1.0.0 - 2021-05-27

* Initial release.
* Configurable hub account, spoke account, and external ID.


cmd_run.go
package main

import (
	"os"

	errors "github.com/go-faster/errors"
	cli "github.com/jawher/mow.cli"
	"github.com/northwood-labs/awsutils"
	"github.com/northwood-labs/golang-utils/exiterrorf"

	"github.com/northwood-labs/assume-spoke-role/hubspoke"
)

func cmdRun(cmd *cli.Cmd) {
	cmd.LongDesc = `Perform the action of assuming roles and running an action.

Use environment variables to store parameter values consistently. CLI options
take precedence over environment variables.`

	const (
		defaultAWSRetries = 3
	)

	var (
		externalID    = os.Getenv("ASSUME_ROLE_EXTERNAL_ID")
		hubAccount    = os.Getenv("ASSUME_ROLE_HUB_ACCOUNT")
		hubRole       = os.Getenv("ASSUME_ROLE_HUB_ROLE")
		spokeAccount  = os.Getenv("ASSUME_ROLE_SPOKE_ACCOUNT")
		spokeRole     = os.Getenv("ASSUME_ROLE_SPOKE_ROLE")
		sessionString = os.Getenv("ASSUME_ROLE_SESSION_STRING")

		externalIDFlag    = cmd.StringOpt("e external-id", externalID, "(ASSUME_ROLE_EXTERNAL_ID) The external ID value that is required by your hub and spoke policies, if any.")   // lint:ignore-length
		hubAccountFlag    = cmd.StringOpt("h hub-account", hubAccount, "(ASSUME_ROLE_HUB_ACCOUNT) The 12-digit AWS account ID containing the HUB policy.")                           // lint:ignore-length
		spokeAccountFlag  = cmd.StringOpt("s spoke-account", spokeAccount, "(ASSUME_ROLE_SPOKE_ACCOUNT) The 12-digit AWS account ID containing the SPOKE policy.")                   // lint:ignore-length
		hubRoleFlag       = cmd.StringOpt("H hub-role", hubRole, "(ASSUME_ROLE_HUB_ROLE) The name of the IAM role to assume in the HUB account.")                                    // lint:ignore-length
		spokeRoleFlag     = cmd.StringOpt("S spoke-role", spokeRole, "(ASSUME_ROLE_SPOKE_ROLE) The name of the IAM role to assume in the HUB account.")                              // lint:ignore-length
		sessionStringFlag = cmd.StringOpt("session-string", sessionString, "(ASSUME_ROLE_SESSION_STRING) A string that will be part of the resulting User ID in the spoke account.") // lint:ignore-length
		retriesFlag       = cmd.IntOpt("r retries", defaultAWSRetries, "The maximum number of retries that the underlying AWS SDK should perform.")                                  // lint:ignore-length
		verboseFlag       = cmd.BoolOpt("v verbose", false, "Enable verbose logging.")                                                                                               // lint:ignore-length
		cmdd              = cmd.StringsArg("COMMAND", []string{""}, "The command to run using the spoke policy.")
	)

	cmd.Spec = `[-e=<external-id>] [-h=<hub-account>] [-s=<spoke-account>] [-H=<hub-role>] ` +
		`[-S=<spoke-role>] [-r=<retries>] [--verbose] -- COMMAND...`

	cmd.Action = func() {
		// Get AWS credentials from environment.
		config, err := awsutils.GetAWSConfig(ctx, "", "", *retriesFlag, *verboseFlag)
		if err != nil {
			exiterrorf.ExitErrorf(errors.Wrap(err, "could not generate a valid AWS configuration object"))
		}

		// Assume appropriate roles and return session credentials for the "Spoke" account.
		roleCredentials, _, err := hubspoke.GetSpokeCredentials(&hubspoke.SpokeCredentialsInput{
			Context:        ctx,
			Config:         &config,
			HubAccountID:   *hubAccountFlag,
			SpokeAccountID: *spokeAccountFlag,
			HubRoleName:    *hubRoleFlag,
			SpokeRoleName:  *spokeRoleFlag,
			ExternalID:     *externalIDFlag,
			SessionString:  *sessionStringFlag,
		})
		if err != nil {
			exiterrorf.ExitErrorf(errors.Wrap(err, "could not generate valid AWS credentials for the 'spoke' account"))
		}

		// Pass the spoke credentials to a CLI task.
		runCommand(roleCredentials, *cmdd)
	}
}


hubspoke/functions.go
package hubspoke

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials/stscreds"
	"github.com/aws/aws-sdk-go-v2/service/sts"
	"github.com/aws/aws-sdk-go-v2/service/sts/types"
	errors "github.com/go-faster/errors"
)

// SpokeCredentialsInput is an input object for the GetSpokeCredentials
// function.
type SpokeCredentialsInput struct {
	// (Required) A context.Context object.
	Context context.Context

	// (Required) An AWS SDK v2 configuration object.
	Config *aws.Config

	// (Required) The AWS Account ID of the "hub" account.
	HubAccountID string

	// (Required) The AWS Account ID of the "spoke" account.
	SpokeAccountID string

	// (Required) The name of the role to assume inside the "hub" account.
	HubRoleName string

	// (Required) The name of the role to assume inside the "spoke" account.
	SpokeRoleName string

	// (Optional) A string identifier that should match what the IAM policies
	// require, if anything. If empty, an empty string will be passed along.
	ExternalID string

	// (Optional) A string identifier to use to represent the
	// user/software/role, and will show up under the `UserId` result of `aws
	// sts get-caller-identity`. If empty, a random string will be generated.
	SessionString string
}

// GetSpokeCredentials accepts a GetSpokeCredentialsInput object, and returns a
// set of STS session credentials for the spoke account.
func GetSpokeCredentials(input *SpokeCredentialsInput) (*types.Credentials, aws.Config, error) {
	const randomStringLength = 32

	emptyCredentials := types.Credentials{}
	emptyConfig := aws.Config{}

	sessionName := input.SessionString
	if sessionName == "" {
		rand.Seed(time.Now().UnixNano())

		sessionName = randomString(randomStringLength)
	}

	hubRoleARN := fmt.Sprintf("arn:aws:iam::%s:role/%s", input.HubAccountID, input.HubRoleName)
	spokeRoleARN := fmt.Sprintf("arn:aws:iam::%s:role/%s", input.SpokeAccountID, input.SpokeRoleName)

	// Assume the HUB role.
	stsHubClient := sts.NewFromConfig(*input.Config)
	input.Config.Credentials = aws.NewCredentialsCache(
		stscreds.NewAssumeRoleProvider(stsHubClient, hubRoleARN, func(o *stscreds.AssumeRoleOptions) {
			o.ExternalID = aws.String(input.ExternalID)
		}),
	)

	// Assume the SPOKE role.
	stsSpokeClient := sts.NewFromConfig(*input.Config)
	response, err := stsSpokeClient.AssumeRole(input.Context, &sts.AssumeRoleInput{
		RoleArn:         aws.String(spokeRoleARN),
		RoleSessionName: aws.String(fmt.Sprintf("%s-%s", input.SpokeAccountID, sessionName)),
		ExternalId:      aws.String(input.ExternalID),
	})
	if err != nil { // lint:allow_cuddling
		return &emptyCredentials, emptyConfig, errors.Wrap(err, fmt.Sprintf(
			"error assuming '%s' role in account %s",
			spokeRoleARN,
			input.SpokeAccountID,
		))
	}

	input.Config.Credentials = aws.NewCredentialsCache(
		stscreds.NewAssumeRoleProvider(stsSpokeClient, spokeRoleARN, func(o *stscreds.AssumeRoleOptions) {
			o.RoleSessionName = fmt.Sprintf("%s-%s", input.SpokeAccountID, sessionName)
			o.ExternalID = aws.String(input.ExternalID)
		}),
	)

	return response.Credentials, *input.Config, nil
}

func randomString(l int) string {
	const (
		// Pool of characters to choose from when generating a random session ID.
		pool = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
	)

	bytes := make([]byte, l)

	for i := 0; i < l; i++ {
		bytes[i] = pool[rand.Intn(len(pool))] // lint:not_crypto
	}

	return string(bytes)
}


scripts/dependency-pip.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

import pathlib
import re
import subprocess
import sys

MIN_PYTHON = (3, 7)
if sys.version_info < MIN_PYTHON:
    sys.exit(
        "[ERROR] This script requires Python {version} or newer.".format(
            version='.'.join([str(n) for n in MIN_PYTHON])
        )
    )

#-------------------------------------------------------------------------------

reqs = subprocess.check_output([sys.executable, '-m', 'pip', 'freeze'])
installed_packages = [r.decode().split('==')[0] for r in reqs.split()]
script_dir = pathlib.Path(__file__).parent.absolute()

# Python packages
with open(f"{script_dir}/requirements.txt") as f:
    for dep in [line for line in f]:
        if re.match(r"^([^(>|<|=)]+)", dep).group(0) not in installed_packages:
            print(
                "Cannot find dependencies. Running `pip install -r scripts/requirements.txt` automatically."
            )
            reqs = subprocess.check_output([
                sys.executable, '-m', 'pip', 'install', '-r', 'scripts/requirements.txt'
            ])

            print("")
            break


scripts/dependency-check.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

import re
import shutil
import subprocess
import sys
from semantic_version import SimpleSpec, Version

MIN_PYTHON = (3, 7)
if sys.version_info < MIN_PYTHON:
    sys.exit(
        "[ERROR] This script requires Python {version} or newer.".format(
            version='.'.join([str(n) for n in MIN_PYTHON])
        )
    )

#-------------------------------------------------------------------------------

def run(command):
    return subprocess.check_output(
        command.split(" "), stderr=subprocess.STDOUT
    ).decode("utf-8")

def meets_version_requirements(needs_to_meet, current_version):
    needs_to_meet = version_range_handler(needs_to_meet)
    spec = SimpleSpec(needs_to_meet)

    return spec.match(Version(current_version))

def discover_version(text):
    re_version = re.compile(r"v?(\d+\.\d+\.\d+)")
    matches = re_version.search(text)

    if matches == None:
        raise ValueError(f"[ERROR] Could not discover a version number in `{text}`.")

    return matches.group(1)

def version_range_handler(version):
    # Remove extra spaces
    version = version.replace(" ", "")

    # Terraform allows ranges like `~> 0.13`.
    m = re.match("^~>\s*(\d+\.\d+)", version)

    # Uses `~>`
    if m is not None:
        starting_version = Version(f"{m.group(1)}.0")
        ending_version = starting_version.next_minor()
        version = f">={str(starting_version)},<{str(ending_version)}"

    return version

def check_all(
    binary_name,
    needs_version,
    get_version,
    installation_message="",
    unmet_version_message=""
):
    if installation_message == "":
        installation_message = f"If you have macOS and Homebrew, you can install it with `brew install {binary_name}`."

    if shutil.which(binary_name) == None:
        print(
            f"[ERROR] The `{binary_name}` binary is not installed. {installation_message}"
        )

        print("")
        sys.exit(1)

    version_str = discover_version(run(get_version))

    if unmet_version_message == "":
        unmet_version_message = "[ERROR] Requires {binary_name} {needs_version}. Version {version_str} appears to be installed."

    if meets_version_requirements(needs_version, version_str) == False:
        print(
            unmet_version_message.format(
                binary_name=binary_name,
                needs_version=needs_version,
                version_str=version_str,
            )
        )

        print("")
        sys.exit(1)

#-------------------------------------------------------------------------------

# nproc installed
if shutil.which("nproc") == None:
    print(
        "[ERROR] The `nproc` binary is not installed. If you have macOS and Homebrew, you can install it with `brew install coreutils`."
    )

    print("")
    sys.exit(1)

# go
check_all(
    binary_name="go",
    needs_version=">= 1.16",
    get_version="go version",
)

# golangci-lint
check_all(
    binary_name="golangci-lint",
    needs_version=">=1.38.0",
    get_version="golangci-lint version",
)


main.go
package main

import (
	"context"
	"fmt"
	"os"
	"runtime"

	errors "github.com/go-faster/errors"
	"github.com/gookit/color"
	cli "github.com/jawher/mow.cli"
	"github.com/northwood-labs/golang-utils/exiterrorf"
)

var (
	// Make referencable throughout.
	app *cli.Cli
	err error

	ctx = context.Background()

	// Color text.
	colorHeader = color.New(color.FgWhite, color.BgBlue, color.OpBold)

	// Buildtime variables.
	commit  string
	date    string
	version string
)

func main() {
	desc := `Assumes a 'spoke' role, from a 'hub' role, from a resource.

Example #1:

    export ASSUME_ROLE_EXTERNAL_ID=this-is-my-robot
    export ASSUME_ROLE_HUB_ACCOUNT=999999999999
    export ASSUME_ROLE_HUB_ROLE=robot-hub-role
    export ASSUME_ROLE_SPOKE_ROLE=robot-spoke-role

    aws-vault exec sys-robot --no-session -- \
        assume-spoke-role --spoke-account 888888888888 -- \
            aws sts get-caller-identity

Example #2:

    aws-vault exec sys-robot --no-session -- \
        assume-spoke-role \
            --hub-account 999999999999 \
            --spoke-account 888888888888 \
            --hub-role robot-hub-role \
            --spoke-role robot-spoke-role \
            --external-id this-is-my-robot \
            -- \
                aws sts get-caller-identity`

	app = cli.App("assume-spoke-role", desc)
	app.Version("version", fmt.Sprintf(
		"assume-spoke-role %s (%s/%s)",
		version,
		runtime.GOOS,
		runtime.GOARCH,
	))

	app.Command("run", "Perform the action of assuming roles and running an action.", cmdRun)
	app.Command("version", "Verbose information about the build.", cmdVersion)

	err = app.Run(os.Args)
	if err != nil {
		exiterrorf.ExitErrorf(errors.Wrap(err, "failed to execute application"))
	}
}
